cip/CIP2015-06-24-call-procedures.adoc

```diff
@@ -0,0 +1,250 @@
+
+[abstract]
+.Abstract
+
+This proposal introduces stored procedures as a new concept to Cypher and proposes a new facility
+for calling them.
+
+User-defined functions are a separate concern that might be addressed in a future CIP.
+
+toc::[]
+
+== Motivation & Background
+
+While Cypher is a very expressive graph query language, its declarative nature and scope is currently not sufficient for
+expressing certain classes of graph traversals, like traversals relying on using specialized data structures or on accessing third-party functionality.
+
+This CIP proposes to add stored procedures to Cypher in order to address these issues.
+
+== Proposal
+
+This proposal consists of two parts, introducing stored procedures conceptually, and describing how to call them in a
+system that supports Cypher.
+
+=== Stored procedures
+
+A stored procedure has a name and takes a fixed number of named and typed arguments in a fixed sequence and either just
+yields nothing (i.e. only potentially causes a side effect) or yields a stream of result records.
+Each record yielded contains values for a fixed number of named and typed fields in the same fixed sequence such that each field value in a yielded record matches its field's type.
+
+The name of a stored procedure, its argument names, and if applicable any result field names follow the same rules as other symbolic names in Cypher (like variables, labels, relationship types, and property keys).
+
+==== Procedure signatures
+
+The arguments of a procedure, their types, and their order are called its argument signature.
+All information about what is being returned by the procedure (either if the procedure yields nothing or the yielded result field names and types, as well as their order) are called the result signature of the procedure.
+Together, the argument signature and the result signature form the (full) signature of a stored procedure.
+
+According to this definition, a procedure may take zero arguments and/or may yield records with zero result fields.
+Note that yielding records with zero result fields is considered different from yielding nothing.
+
+==== Procedure signature notation
+
+This CIP suggests a standard notation for printing procedure names together with their signatures.
+This notation may be used in system monitoring, documentation and error reporting, and perhaps more importantly is to be used in future CIPs.
+It is based on the syntax for type annotations as specified by the Cypher type system:
+
+- Procedures returning nothing: `<procName>(arguments) :: VOID`
+- Procedures returning a stream of records: `<procName>(arguments) :: (results)`
+- Arguments and results are written as comma-separated lists of `<name> :: <type>`
+
+An actual implementation may choose to support a notion of default value for procedure arguments.
+For such a case, this CIP suggests that arguments with default values are written as `<name> = <value> :: <type`.
```

Comment:
I don't think  we should have "default values" be something that is left to the implementation. We should just specify how default values are notated.